OrbitCore/LinuxEventTracer.cpp

```diff
@@ -0,0 +1,262 @@
+{
+    //m_Callback = a_Callback ? a_Callback : [this](const std::string& a_Buffer)
+    //{
+    //    CommandCallback(a_Buffer);
+    //};
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Start()
+{
+    PRINT_FUNC;
+#if __linux__
+    m_ExitRequested = false;
+
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::foo, &m_ExitRequested);
+
+   /* m_Thread = std::make_shared<std::thread>
+        ( &LinuxUtils::StreamCommandOutput
+        , "perf record -e sched:sched_switch -o - | perf script -i -"
+        , m_Callback
+        , &m_ExitRequested );
+    */
+    m_Thread->detach();
+#endif
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Stop()
+{
+    m_ExitRequested = true;
+}
+
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::CommandCallback(std::string a_Line)
+{
+    std::vector<std::string> tokens = Tokenize(a_Line);
+    
+    // TODO: log broken lines
+    if (tokens[4] != "sched:sched_switch:" || tokens[8] != "==>")
+    {
+        return;
+    }
+
+    auto oldTID = atoi(tokens[1].c_str());
+    auto cpu = atoi(trim(tokens[2], "[]").c_str());
+    auto time = GetMicros(tokens[3])*1000;
+    char state = tokens[7].at(0);
+
+    auto newCommCPUTID = Tokenize(tokens[9], "/:");
+    auto newTID = atoi(newCommCPUTID.back().c_str());
+
+    // TODO: only process context switches that relate to the traced process.
+
+    ContextSwitch::SwitchType type = ContextSwitch::Invalid;
+    int tid = -1;
+    switch (state)
+    {
+    // sleep:
+    case 'S':
+        //if (!Capture::GTargetProcess->HasThread(oldTID))
+        //    return;
+
+        type = ContextSwitch::Out;
+        tid = oldTID;
+        break;
+    
+    // run:
+    case 'R':
+        //if (!Capture::GTargetProcess->HasThread(newTID))
+        //    return;
+
+        type = ContextSwitch::In;
+        tid = newTID;
+        break;
+
+    // TODO: handle the other cases
+
+    default:
+        return;
+    }
+
+    if (state != ContextSwitch::Invalid && tid != -1)
+    {
+        ++Capture::GNumContextSwitches;
+
+        ContextSwitch CS ( type );
+        CS.m_ThreadId = tid;
+        CS.m_Time = time;
+        CS.m_ProcessorIndex = cpu;
+        //TODO: Is this correct?
+        CS.m_ProcessorNumber = cpu;
+        GTimerManager->Add( CS );
+    }
+}
+
+void LinuxEventTracer::foo(bool* a_ExitRequested)
+{
+    int cpus = std::stoi(LinuxUtils::ExecuteCommand("nproc"));
+    uint64_t tracepointID = LinuxUtils::GetTracePointID("sched", "sched_switch");
+    std::vector<uint32_t> fds;
+
+    // open perf events for all cpus to keep track of process spawning
+    for (int cpu = 0; cpu < cpus; cpu++)
+    {
+        fds.push_back(LinuxPerfUtils::task_event_open(cpu)); 
+        fds.push_back(LinuxPerfUtils::tracepoint_event_open(tracepointID, -1 /*pid*/, cpu)); 
+    }
+
```

Comment:
Remove one of the empty lines here? Not sure what the code style here should be.